subsys/bluetooth/controller/ll_sw/ull_conn.c

```diff
@@ -3157,35 +3182,18 @@ static inline void event_len_prep(struct ll_conn *conn)
 		lr->max_tx_octets = sys_cpu_to_le16(conn->default_tx_octets);
 
 		if (!conn->common.fex_valid ||
-#if defined(CONFIG_BT_CTLR_PHY)
-		     (
-#if defined(CONFIG_BT_CTLR_PHY_CODED)
-		      !(conn->llcp_feature.features &
-			BIT(BT_LE_FEAT_BIT_PHY_CODED)) &&
-#endif /* CONFIG_BT_CTLR_PHY_CODED */
-
-#if defined(CONFIG_BT_CTLR_PHY_2M)
-		      !(conn->llcp_feature.features &
-			BIT(BT_LE_FEAT_BIT_PHY_2M)) &&
-#endif /* CONFIG_BT_CTLR_PHY_2M */
-		      1)
-#else /* !CONFIG_BT_CTLR_PHY */
-		    0
-#endif /* !CONFIG_BT_CTLR_PHY */
-		   ) {
+		    (!feature_coded_phy && !feature_phy_2m)) {
```

Comment:
LHS evaluates to 1 when both coded PHY and 2M is not supported by local device (when only the procedure is supported). In the new code (RHS), it evaluates to 0. OR-ing with 0 means fex_valid will be checked and if there was prior feature exchange then max_/rx/tx_time fields will not be filled  as else-clauses have been compiled-out.

Reply:
You are talking about the case when all three CONFIG_BT_CTLR_PHY, CONFIG_BT_CTLR_PHY_CODED and CONFGI_BT_CTLR_PHY_2M are valid?

In that case the original code, when removing the conditional compiles is:
		if (!conn->common.fex_valid ||
		     (
		      !(conn->llcp_feature.features &
			BIT(BT_LE_FEAT_BIT_PHY_CODED)) &&
		      !(conn->llcp_feature.features &
			BIT(BT_LE_FEAT_BIT_PHY_2M))
		   )

Replacing the 'conn.>llcp_feature.features & BIT(x))' with 'feature_x' gives the code as shown

My understanding of what should happen is:
if (we had a feature exchange) or (both encoded phy and 2mbit phy invalid)
then
... do something
else if (encoded phy)
...
else if (2mbit phy)
...
fi

Reply:
Added a commit to fix the obscure bug; refactored code should take care of this bug